From 2543c4cc973027f9423cab3274f22917228d56d4 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:01 -0400
Subject: [fs] nfsd: Fix bad checking of space for padding in splice read

Message-id: <1410898849-22126-58-git-send-email-bfields@redhat.com>
Patchwork-id: 95814
O-Subject: [Patch RHEL7 057/165] NFSD: Fix bad checking of space for padding in splice read
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

Note that the caller has already reserved space for count and eof, so
xdr->p has already moved past them, only the padding remains.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Fixes dc97618ddd (nfsd4: separate splice and readv cases)
Signed-off-by: J. Bruce Fields <bfields@redhat.com>

Upstream: d5d5c304b13bc3cade13b8a1b5833c8b3a0975f1
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index f850929..08677a2 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -3072,11 +3072,8 @@ static __be32 nfsd4_encode_splice_read(
  __be32 nfserr;
  __be32 *p = xdr->p - 2;
 
- /*
-  * Don't inline pages unless we know there's room for eof,
-  * count, and possible padding:
-  */
- if (xdr->end - xdr->p < 3)
+ /* Make sure there will be room for padding if needed */
+ if (xdr->end - xdr->p < 1)
   return nfserr_resource;
 
  nfserr = nfsd_splice_read(read->rd_rqstp, file,
-- 
1.7.1